### PR TITLE
[New]`jsx-newline`: Enforce a new line after jsx elements and expressions

### LINKED ISSUE #2630: Help wanted: Spacing between tags
Hi everybody, I am searching for hours and I really can't find a solution to my problem. I will appreciate if somebody can help me with this. 
My problem is related to tag formatting and spacing. 
I need a rule that will enforce adding **newlines between siblings**.

Bad:
```javascript
<div>
  <Button>{data.label}</Button>
  <List />
  <Button>
    <IconPreview /> 
    Button 2
  </Button>
  {showSomething === true ? (
    <Something />
  )}
  <Button>Button 3</Button>
  {showSomething === true ? (
    <Something />
  ) : (
    <ErrorMessage />
  )}
</div>
```

Good:
```javascript
<div>
  <Button>{data.label}</Button>

  <List />

  <Button>
    <IconPreview /> 
    Button 2
  </Button>

  {showSomething === true ? (
    <Something />
  )}

  <Button>Button 3</Button>

  {showSomethingElse === true ? (
    <SomethingElse />
  ) : (
    <ErrorMessage />
  )}
</div>
```

I search in the prettier's rules, eslint's rules, everywhere and I am not finding it. If anybody knows a rule like this or point me to something I will be grateful.

### PR DESCRIPTION
**Pull request by jzabala**
Closes #2630

### CODE REVIEW

**Comment by jzabala** [comment]
I don't get why the README generation of the build is failing. I need help.

**Comment by ljharb** [comment]
@jzabala the autogeneration isn't failing; you have to manually add it, and the autogeneration *validates* that you added it.

File: docs/rules/jsx-newline.md [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+```jsx
+<div>
+  <Button>{data.label}</Button>
+
+  <List />
+
+  <Button>
+    <IconPreview />
+    Button 2
```
Comment (ljharb) [inline docs/rules/jsx-newline.md:46]: how is this not a warning? there's no newline after the element.

File: index.js [hunk 2]
```diff
@@ -91,7 +91,8 @@ const allRules = {
   'state-in-constructor': require('./lib/rules/state-in-constructor'),
   'static-property-placement': require('./lib/rules/static-property-placement'),
   'style-prop-object': require('./lib/rules/style-prop-object'),
-  'void-dom-elements-no-children': require('./lib/rules/void-dom-elements-no-children')
+  'void-dom-elements-no-children': require('./lib/rules/void-dom-elements-no-children'),
+  'jsx-newline': require('./lib/rules/jsx-newline')
```
Comment (ljharb) [inline index.js:95]: this list is alphabetized

File: lib/rules/jsx-newline.js [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    fixable: 'code'
+  },
+  create(context) {
+    const jsxElementParents = new Set();
+    const sourceCode = context.getSourceCode();
+    return {
+      'Program:exit'() {
+        jsxElementParents.forEach((parent) => {
+          parent.children.forEach((element, index, elements) => {
+            if (arrayIncludes(['JSXElement', 'JSXExpressionContainer'], element.type)) {
```
Comment (ljharb) [inline lib/rules/jsx-newline.js:32]:
```suggestion
            if (element.type === 'JSXElement' || element.type === 'JSXExpressionContainer') {
```
no need to make an array for just 2 items :-)

File: lib/rules/jsx-newline.js [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+              const firstAdjacentSibling = elements[index + 1];
+              const secondAdjacentSibling = elements[index + 2];
+              if (
+                firstAdjacentSibling
+                && secondAdjacentSibling
+                && arrayIncludes(['Literal', 'JSXText'], firstAdjacentSibling.type)
```
Comment (ljharb) [inline lib/rules/jsx-newline.js:38]: same here

File: docs/rules/jsx-newline.md (hunk 1 above)
Comment (jzabala) [reply docs/rules/jsx-newline.md:46]: In the **Good** part of the issue description the same example is giving without a newline. That is why the rule only warns about newlines between elements and expressions, no literals like text and numbers.

File: index.js (hunk 2 above)
Comment (jzabala) [reply index.js]: Solved!

File: lib/rules/jsx-newline.js (hunk 3 above)
Comment (jzabala) [reply lib/rules/jsx-newline.js]: Solved!

File: lib/rules/jsx-newline.js (hunk 4 above)
Comment (jzabala) [reply lib/rules/jsx-newline.js]: Solved!

**Comment by jzabala** [comment]
> @jzabala the autogeneration isn't failing; you have to manually add it, and the autogeneration validates that you added it.
Fixed. Thanks.

> Let's make sure all the test cases are ran in the default parser, babel-eslint, and the typescript parser.
Done. 
Note: the tests with Fragments are not run against the default parser. Those were breaking the build.

Hi @ljharb. I addressed all your comments. Thanks for taking the time to review the PR 🙂

**Comment by ljharb** [comment]
@rumenpetrov can you confirm that this rule will meet your needs?

**Comment by rumenpetrov** [comment]
@jzabala @ljharb thank you for your great work. The PR almost covers my needs. I just have a remark about the case where we have text along with other components. I didn't provide enough details last time on how it should be implemented, so this is on me, sorry about that. I will try to clear it now.

When there is a text as a child of a parent component I don't need the new lines for its children. I need the new lines only when the children are other components and there is no text. 
Let me know if this makes sense to you or if you think this could be a potential problem.

**Bad**
``` 
<h1>
  This is
  <em>short</em>

  <strong> title</strong>
</h1>

// The new line before the span here could be a potential conflict with 'prettier'
<Button>
  <IconPreview />
  Really looooooooooooooong button name

  <span>a suffix</span>
</Button>
```

**Good**
```
<h1>This is <em>short</em><strong> title</strong></h1>

<Button>
  <IconPreview />
  Really looooooooooooooong button name
  <span>a suffix</span>
</Button>
```

Also, I am not really sure how to test this PR locally or with a real project. @jzabala can you point me to a documentation page or an article related to this?

**Comment by doumart** [comment]
Don't want to annoy anyone but does somebody know if PR is still in the process of being merged?

**Review by ljharb — APPROVED** [review]

**Pull request merged**